File: USACOTrain/old3_1/agrinet.java
package old3_1;/*
ID: derekl3
LANG: JAVA
TASK: old3_1.agrinet
*/

import java.io.*;
import java.util.*;

public class agrinet {
	static int INF = 1000000;
	public static void main(String[] args) throws IOException {
		BufferedReader fin = new BufferedReader(new FileReader("old3_1.agrinet.in"));
		PrintWriter fout =  new PrintWriter(new FileWriter("old3_1.agrinet.out"));
		int F = Integer.parseInt(fin.readLine());
		StringTokenizer sT;
		//assemble graph
		int [][] G = new int[F][F];
		for (int i = 0; i < F; i++) {
			sT = new StringTokenizer(fin.readLine());
			for (int j = 0; j < F; j++) {
				if (!sT.hasMoreTokens()) sT = new StringTokenizer(fin.readLine());
				G[i][j] = Integer.parseInt(sT.nextToken());
			}
		}
		fin.close();
		//set up tree
		int[][] T = new int[F][3]; //distance, intree, source
		for (int[] k : T) k[0] = INF;
		//use Prim's algorithm
		int tS = 1; //treeSize
		int tC = 0; //treeCost
		T[0][1] = 1; // adds node 0 to tree
		for (int l = 0; l < F; l++){
			T[l][0] = G[0][l]; //update neighbors in tree	
		}
		while (tS < F) {
			//check for closest node to tree
			int minNode = INF;
			int minDist = INF;
			for (int m = 0; m < F; m++) {
				if (T[m][0] < minDist && T[m][1] == 0) {
					minDist = T[m][0];
					minNode = m;
				}
			}
			tS++;
			tC += minDist;
			T[minNode][1] = 1;
			for (int n = 0; n < F; n++) {
				//update nodes connected to minNode
				if (G[minNode][n] < T[n][0]) {
					T[n][0] = G[minNode][n];
					T[n][2] = minNode;
				}
			}
		}
		System.out.println(tC);
		fout.println(tC);
		fout.close();
	}
}

File: USACOTrainCPP/agrinet.cpp
/*
ID: derek.l3
TASK: agrinet
LANG: C++
*/
//USACO Training 3.1: Agrinet

#include <iostream>
#include <fstream>

using namespace std;

int N;
int C[101][101];
int dist[101];
int source[101];
bool V[101];
int main () {
	int i, j;
	int ts=0, tc=0, cur=1;

	ifstream fin("agrinet.in");
	ofstream fout("agrinet.out");

	fin >> N;
	for (i=1;i<=N;i++)
	{
		dist[i]=100000;
		source[i]=1;
		for (j=1;j<=N;j++)
		{
			fin >> C[i][j];
		}
	}
	
	while (ts!=N)
	{
		for (i=1;i<=N;i++)
		{
			if ((V[cur] || dist[i]<dist[cur]) && !V[i])
			{
				cur=i;
			}
		}
		ts++;
		tc+=C[source[cur]][cur];
		V[cur]=true;

		for (i=1;i<=N;i++)
		{
			if (C[cur][i]<=dist[i]) 
			{
				source[i]=cur;
				dist[i]=C[cur][i];
			}
		}
	}
	cout << tc << endl;
	fout << tc << endl;
	return 0;
}
